syz-ci/jobs.go

```diff
@@ -26,72 +27,146 @@ import (
 	"github.com/google/syzkaller/vm"
 )
 
+type JobManager struct {
+	cfg             *Config
+	dash            *dashapi.Dashboard
+	managers        []*Manager
+	shutdownPending <-chan struct{}
+}
+
 type JobProcessor struct {
 	cfg             *Config
 	name            string
+	instanceSuffix  string
 	managers        []*Manager
 	knownCommits    map[string]bool
-	stop            chan struct{}
-	shutdownPending chan struct{}
 	dash            *dashapi.Dashboard
-	syzkallerRepo   string
-	syzkallerBranch string
+	baseDir         string
+	jobFilter       *ManagerJobs
+	jobTicker       <-chan time.Time
+	commitTicker    <-chan time.Time
+	shutdownPending <-chan struct{}
 }
 
-func newJobProcessor(cfg *Config, managers []*Manager, stop, shutdownPending chan struct{}) (*JobProcessor, error) {
+func newJobManager(cfg *Config, managers []*Manager, shutdownPending chan struct{}) (*JobManager, error) {
 	dash, err := dashapi.New(cfg.DashboardClient, cfg.DashboardAddr, cfg.DashboardKey)
 	if err != nil {
 		return nil, err
 	}
-	return &JobProcessor{
+	return &JobManager{
 		cfg:             cfg,
-		name:            fmt.Sprintf("%v-job", cfg.Name),
+		dash:            dash,
 		managers:        managers,
-		knownCommits:    make(map[string]bool),
-		stop:            stop,
 		shutdownPending: shutdownPending,
-		dash:            dash,
-		syzkallerRepo:   cfg.SyzkallerRepo,
-		syzkallerBranch: cfg.SyzkallerBranch,
 	}, nil
 }
 
-func (jp *JobProcessor) loop() {
-	jobTicker := time.NewTicker(time.Duration(jp.cfg.JobPollPeriod) * time.Second)
-	commitTicker := time.NewTicker(time.Duration(jp.cfg.CommitPollPeriod) * time.Second)
-	defer jobTicker.Stop()
+// backgroundLoop starts a job loop in parallel and returns a blocking function
+// to gracefully stop job processing.
+func (jm *JobManager) backgroundLoop(wg *sync.WaitGroup) func() {
+	stop := make(chan struct{})
+	done := make(chan struct{}, 1)
+	go func() {
+		jm.loop(stop)
+		done <- struct{}{}
+		wg.Done()
+	}()
+	return func() {
+		stop <- struct{}{}
+		<-done
+	}
+}
+
+func (jm *JobManager) loop(stop chan struct{}) {
+	if err := jm.resetJobs(); err != nil {
+		if jm.dash != nil {
+			jm.dash.LogError("syz-ci", "reset jobs failed: %v", err)
+		}
+		return
+	}
+	commitTicker := time.NewTicker(time.Duration(jm.cfg.CommitPollPeriod) * time.Second)
 	defer commitTicker.Stop()
+	jobTicker := time.NewTicker(time.Duration(jm.cfg.JobPollPeriod) * time.Second)
+	defer jobTicker.Stop()
+	var wg sync.WaitGroup
+	for parallel := false; ; parallel = true {
+		jp := &JobProcessor{
+			cfg:             jm.cfg,
+			name:            fmt.Sprintf("%v-job", jm.cfg.Name),
+			instanceSuffix:  "-job",
+			managers:        jm.managers,
+			commitTicker:    commitTicker.C,
+			jobTicker:       jobTicker.C,
+			baseDir:         osutil.Abs("jobs"),
+			knownCommits:    make(map[string]bool),
+			dash:            jm.dash,
+			shutdownPending: jm.shutdownPending,
+		}
+		if parallel {
```

Comment:
I find it hard to follow that we initialize part of the fields, but then overwrite and reset a large subset of them.
I think it will be clear to explicitly express what's common and what's different for the 2 processors:
```
		jp := &JobProcessor{
			JobManager: jm,  // embed this completely since we need all fields
			jobTicker:  time.NewTicker(time.Duration(jm.cfg.JobPollPeriod) * time.Second),
		}
		if main {
			jp.instanceSuffix = "jobs"
			jp.commitTicker = time.NewTicker(time.Duration(jm.cfg.CommitPollPeriod) * time.Second).C
			jp.knownCommits = make(map[string]bool)
		} else {
			// For now let's only parallelize patch testing requests.
			jp.instanceSuffix = "jobs-parallel"
			jp.jobFilter = &ManagerJobs{TestPatches: true}
		}
		jp.name = fmt.Sprintf("%v-%v", jm.cfg.Name, jp.instanceSuffix),
		jp.baseDir = osutil.Abs(jp.instanceSuffix)
```

Reply:
Thanks!
I've restructured the code like this, though with one exception -- I think it'd be better not to change `jobSuffix` and `baseDir` for our main processor, it's just one-two more lines, but it will let us have much less leftover garbage in our GCE project and on our syz-ci instances.